### PR TITLE
Queue async operations on a per-player basis [WIP]

### PR DESCRIPTION
**Pull request by sgdc3**

### CODE REVIEW

**Comment by sgdc3** [comment]
Depends on #1992

File: src/main/java/fr/xephi/authme/process/AsyncUserScheduler.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+        private Map<K, SequentialExecutor> executors;
+        private Function<Runnable, BukkitTask> scheduler;
+
+        public KeyedSequentialScheduler(Function<Runnable, BukkitTask> scheduler) {
+            this.executors = new LinkedHashMap<>();
+            this.scheduler = scheduler;
+        }
+
+        public void submit(K key, Runnable runnable) {
+            executors.computeIfAbsent(key, k -> new SequentialExecutor(scheduler, () -> executors.remove(key)))
```
Comment (games647) [inline src/main/java/fr/xephi/authme/process/AsyncUserScheduler.java]: Potential memory leak

File: src/main/java/fr/xephi/authme/process/AsyncUserScheduler.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+        private BukkitTask executor;
+
+        public SequentialExecutor(Function<Runnable, BukkitTask> scheduler, Runnable callback) {
+            this.queue = new LinkedBlockingQueue<>();
+            this.scheduler = scheduler;
+            this.callback = callback;
+        }
+
+        public void submit(Runnable task) {
+            queue.add(task);
```
Comment (games647) [inline src/main/java/fr/xephi/authme/process/AsyncUserScheduler.java]: This doesn't seem thread-safe. Considering two calls to the `submit` method. Then both tasks will be added, but `scheduler.apply()` will be invoked twice with two new tasks on the Bukkit scheduler resulting to two new threads. 

T1 and T2 are the threads:
T1: 1 iteration of the while loop
-- 1 task remaining
T1: evaluate `while(!queue.isEmpty())` -> `true` - enter `while` loop
T2: 1 iteration of the while loop
-- `queue.poll` wasn't executed yet by T1 and so `queue` is still not empty
-- T2 now removes the last task
T1: `queue.poll()` - `NoSuchElementExeception`

**Review by TuxCoding — CHANGES_REQUESTED** [review]
I kind of understand what you are doing here. You are trying to perform asynchronous tasks sequential on per player basis. I guess with the goal to reduce conflicts between them? Do you also thought about locks? This allows more flexibility like read/write differentiation, protecting only thread-unsafe operations meanwhile other parts could execute in `concurrently` and in `parallel`.

**Comment by sgdc3** [comment]
@games647 i think that both read/write operations needs to be sequential, I can't think about any situation where locks might be better. Can you make some examples?

File: src/main/java/fr/xephi/authme/process/AsyncUserScheduler.java (hunk 2 above)
Comment (sgdc3) [reply src/main/java/fr/xephi/authme/process/AsyncUserScheduler.java]: Hmm I see, which pattern should we use instead?

Comment (games647) [reply src/main/java/fr/xephi/authme/process/AsyncUserScheduler.java]: The easiest solution would be to lock on the queue on `isEmpty` + `poll` operations.

**Pull request closed without merging**